gokey.go

```diff
@@ -58,6 +62,60 @@ func GetRaw(password, realm string, seed []byte, allowUnsafe bool) (io.Reader, e
 	return rng, nil
 }
 
+// below code implements asn1 encoding of x25519 and ed25519 keys according
+// to https://tools.ietf.org/id/draft-ietf-curdle-pkix-10.txt
+// the output should be compatible to OpenSSL pkey functions
+// this code is considered temporal and is expected to go away, when Go
+// implements native marshalling for these types of keys
+// as https://tools.ietf.org/id/draft-ietf-curdle-pkix-10.txt is still a draft
+// future versions of gokey may produce different output
+
+// p.3 https://tools.ietf.org/id/draft-ietf-curdle-pkix-10.txt
+// id-X25519    OBJECT IDENTIFIER ::= { 1 3 101 110 }
+// id-Ed25519   OBJECT IDENTIFIER ::= { 1 3 101 112 }
+const (
+	x25519OidSuffix  = 110
+	ed25519OidSuffix = 112
+)
+
+// x25519/ed25519 asn1 private key structure
+// p.7 https://tools.ietf.org/id/draft-ietf-curdle-pkix-10.txt
+// this implementation does not support optional attributes or public key
+type asn25519 struct {
+	Version    int
+	AlgId      pkix.AlgorithmIdentifier
+	PrivateKey []byte
+}
+
+// Golang does not have a declaration for x25519 keys
+type x25519PrivateKey []byte
+
+func marshal25519PrivateKey(key crypto.PrivateKey) ([]byte, error) {
+	var a25519 asn25519
+	a25519.Version = 0
+	var keyBytes []byte
+
+	switch key.(type) {
+	case x25519PrivateKey:
+		a25519.AlgId = pkix.AlgorithmIdentifier{asn1.ObjectIdentifier{1, 3, 101, x25519OidSuffix}, asn1.RawValue{}}
+		keyBytes = key.(x25519PrivateKey)
+	case *ed25519.PrivateKey:
+		a25519.AlgId = pkix.AlgorithmIdentifier{asn1.ObjectIdentifier{1, 3, 101, ed25519OidSuffix}, asn1.RawValue{}}
+		keyBytes = key.(*ed25519.PrivateKey).Seed()
+	}
+
+	// actual key bytes are double wrapped in octet strings
+	// see p.7 https://tools.ietf.org/id/draft-ietf-curdle-pkix-10.txt
+	privKeyOctetString, err := asn1.Marshal(keyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	a25519.PrivateKey = privKeyOctetString
```

Comment:
in C I would zeroize privKeyOctetString to make sure it doesn't contain sensitive data anymore. not sure how it works in Go

Reply:
No it is not. I've been concerned about it as well in the past, but AFAIK so far this is a lost battle. Even core go crypto key types do not support it.